homeassistant/components/history_stats/data.py

```diff
@@ -223,3 +233,18 @@ def _async_compute_seconds_and_changes(
         # Save value in seconds
         seconds_matched = elapsed
         return seconds_matched, match_count
+
+    def _prune_history_cache(self, start_timestamp: float) -> None:
+        """Remove unnecessary old data from the history state cache from previous runs.
+
+        Update the timestamp of the last record from before the start to the current start time.
+        """
+        trim_count = 0
+        for i, history_state in enumerate(self._history_current_period):
+            if history_state.last_changed < start_timestamp:
+                history_state.last_changed = start_timestamp
+                if i > 0:
+                    trim_count += 1
+            else:
+                break
+        self._history_current_period = self._history_current_period[trim_count:]
```

Comment:
```suggestion
        for i, history_state in enumerate(self._history_current_period):
            if history_state.last_changed >= start_timestamp:
                break
            history_state.last_changed = start_timestamp
            if i > 0:
                trim_count += 1
        if trim_count: # Don't slice if no data was removed
            self._history_current_period = self._history_current_period[trim_count:]
```
Maybe this is a bit cleaner. Please double check my refactoring here as I've had little sleep last night.